src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationService.cs

```diff
@@ -0,0 +1,131 @@
+    private bool Initialized => _initialized;
+
+    /// <summary>
+    /// Authenticate the user of a request by making a request to a remote app using configured
+    /// headers and/or cookies from the current request.
+    /// </summary>
+    /// <param name="originalRequest">The HTTP request to authenticate.</param>
+    /// <param name="cancellationToken">A cancellation token.</param>
+    /// <returns>
+    /// An authenticaiton result including the claims principal of the authenticated user (if any)
+    /// and a status code and headers to include in the response to the request in case the user could not be authenticated.
+    /// </returns>
+    /// <exception cref="InvalidOperationException">
+    /// An invalid operation exception is thrown if
+    /// AuthenticateAsync is called without calling Initialize.
+    /// </exception>
+    public async Task<RemoteAuthenticationResult> AuthenticateAsync(HttpRequest originalRequest, CancellationToken cancellationToken)
+    {
+        if (!Initialized)
+        {
+            _logger.LogError("Remote authentication handler must be initialized before authenticating");
+            throw new InvalidOperationException("Remote authentication handler must be initialized before authenticating");
+        }
+
+        // Create a new HTTP request, but propagate along configured headers or cookies
+        // that may matter for authentication
+        using var authRequest = new HttpRequestMessage();
+        AddHeaders(_options.HeadersToForward, originalRequest, authRequest);
+        AddCookies(_options.CookiesToForward, originalRequest, authRequest);
+
+        // Get the response from the remote app and convert the response into a remote authentication result
+        using var response = await _client.SendAsync(authRequest, cancellationToken);
+        _logger.LogDebug("Received remote authentication response with status code {StatusCode}", response.StatusCode);
+
+        return await _resultFactory.CreateRemoteAuthenticationResultAsync(response, _options);
+    }
+
+    // Add configured headers to the request, or all headers if none in particualr are specified
+    private static void AddHeaders(IEnumerable<string> headersToForward, HttpRequest originalRequest, HttpRequestMessage authRequest)
+    {
+        IEnumerable<string> headerNames = originalRequest.Headers.Keys;
+        if (headersToForward.Any())
+        {
+            headerNames = headerNames.Where(headersToForward.Contains);
+        }
+
+        foreach (var headerName in headerNames)
+        {
+            authRequest.Headers.Add(headerName, originalRequest.Headers[headerName].ToArray());
+        }
+    }
+
+    // Add configured cookies to the request, or all cookies if none in particualr are specified
+    private static void AddCookies(IEnumerable<string> cookiesToForward, HttpRequest originalRequest, HttpRequestMessage authRequest)
+    {
+        IEnumerable<string> cookieNames = originalRequest.Cookies.Keys;
```

Comment:
Why filter and re-build the cookies? It would be simpler and more efficient to forward the whole Cookie header using the header forwarding logic.

Reply:
I was thinking that some users may want to forward only specific cookies (auth-related ones). I'm fine pulling this out, though, and just treating the cookie header as another header to forward. We can always revisit if we hear feedback from customers that they need more control.

Reply:
I've removed this and added "Cookie" to the list of request headers forwarded by default. If we find users need the ability to forward some cookies but not others, we can revisit this.